### DISCUSSION TITLE
Do i need to use sync once when create client?

### DISCUSSION THREAD

**Discussion by nicoxiang** [go-resty/resty#542] [Q&A] [answered]
Do i need to use sync once when create client?
Is that ok to new a client every time when i want to make a api all?

```
var clientOnce struct {
	sync.Once
	c *resty.Client
}

func getClient() *resty.Client {
	clientOnce.Do(func() {
		clientOnce.c = resty.New().
			SetRetryCount(1).
			SetTimeout(5*time.Second).
			SetHeader("h1", "v1").
			SetHeader("h2", "v2").
	})
	return clientOnce.c
}

```

**Comment by moorereason** [comment, marked as answer]
No. You only need one instance of the client which be used to create any number of requests. Each new request will use the settings from the client it's made with.

**Reply by nicoxiang** [reply]
Thanks for your explanation. BTW in the code sample above, I use sync.once to create one singleton client, is this a good way?

**Reply by moorereason** [reply]
Seems reasonable if you're going to have multiple go-routines calling `getClient`.  If you don't need to protect `getClient` from concurrency, a simple nil check would be sufficient.
